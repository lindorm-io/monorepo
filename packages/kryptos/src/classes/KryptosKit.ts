import { KryptosError } from "../errors";
import {
  IKryptos,
  IKryptosEc,
  IKryptosOct,
  IKryptosOkp,
  IKryptosRsa,
} from "../interfaces";
import {
  KryptosAuto,
  KryptosFormat,
  KryptosFrom,
  KryptosFromBuffer,
  KryptosFromJwk,
  KryptosFromString,
  KryptosLike,
  KryptosMakeEcEnc,
  KryptosMakeEcSig,
  KryptosMakeOctEnc,
  KryptosMakeOctSig,
  KryptosMakeOkpEnc,
  KryptosMakeOkpSig,
  KryptosMakeRsaEnc,
  KryptosMakeRsaSig,
} from "../types";
import { KryptosMake } from "../types/private";
import {
  autoMakeConfig,
  calculateKeyOps,
  fromOptions,
  isB64,
  isDer,
  isJwk,
  isPem,
  makeKey,
} from "../utils/private";
import { Kryptos } from "./Kryptos";

type From = {
  auto(options: KryptosFrom): IKryptos;
  b64(options: KryptosFromString): IKryptos;
  der(options: KryptosFromBuffer): IKryptos;
  jwk(options: KryptosFromJwk): IKryptos;
  pem(options: KryptosFromString): IKryptos;
  utf(options: KryptosFromString): IKryptos;
};

type Encryption = {
  ec(options: KryptosMakeEcEnc): IKryptos;
  oct(options: KryptosMakeOctEnc): IKryptos;
  okp(options: KryptosMakeOkpEnc): IKryptos;
  rsa(options: KryptosMakeRsaEnc): IKryptos;
};

type Signature = {
  ec(options: KryptosMakeEcSig): IKryptos;
  oct(options: KryptosMakeOctSig): IKryptos;
  okp(options: KryptosMakeOkpSig): IKryptos;
  rsa(options: KryptosMakeRsaSig): IKryptos;
};

type Make = {
  auto(options: KryptosAuto): IKryptos;
  enc: Encryption;
  sig: Signature;
};

export class KryptosKit {
  // clone

  public static clone(kryptos: IKryptos): IKryptos {
    return new Kryptos({ ...kryptos.toJSON(), ...kryptos.export("der") });
  }

  // from

  public static get from(): From {
    return {
      auto: this.fromAuto,
      b64: this.fromB64,
      der: this.fromDer,
      jwk: this.fromJwk,
      pem: this.fromPem,
      utf: this.fromUtf,
    };
  }

  // is

  public static isEc(kryptos: KryptosLike): kryptos is IKryptosEc {
    return (
      kryptos instanceof Kryptos && kryptos.type === "EC" && kryptos.curve !== undefined
    );
  }

  public static isOct(kryptos: KryptosLike): kryptos is IKryptosOct {
    return (
      kryptos instanceof Kryptos && kryptos.type === "oct" && kryptos.curve === undefined
    );
  }

  public static isOkp(kryptos: KryptosLike): kryptos is IKryptosOkp {
    return (
      kryptos instanceof Kryptos && kryptos.type === "OKP" && kryptos.curve !== undefined
    );
  }

  public static isRsa(kryptos: KryptosLike): kryptos is IKryptosRsa {
    return (
      kryptos instanceof Kryptos && kryptos.type === "RSA" && kryptos.curve === undefined
    );
  }

  // make

  public static get make(): Make {
    return {
      auto: this.makeAuto,
      enc: this.encryption,
      sig: this.signature,
    };
  }

  // private from

  private static fromAuto(options: KryptosFrom): IKryptos {
    if (isB64(options)) return KryptosKit.fromB64(options);
    if (isDer(options)) return KryptosKit.fromDer(options);
    if (isJwk(options)) return KryptosKit.fromJwk(options);
    if (isPem(options)) return KryptosKit.fromPem(options);

    throw new KryptosError("Unexpected key format");
  }

  private static fromB64(options: KryptosFromString): IKryptos {
    return KryptosKit.fromKryptos("b64", options);
  }

  private static fromDer(options: KryptosFromBuffer): IKryptos {
    return KryptosKit.fromKryptos("der", options);
  }

  private static fromJwk(options: KryptosFromJwk): IKryptos {
    return KryptosKit.fromKryptos("jwk", options);
  }

  private static fromPem(options: KryptosFromString): IKryptos {
    return KryptosKit.fromKryptos("pem", options);
  }

  private static fromUtf(options: KryptosFromString): IKryptos {
    return KryptosKit.fromKryptos("utf", options);
  }

  private static fromKryptos(format: KryptosFormat, arg: KryptosFrom): Kryptos {
    const options = fromOptions(format, arg);

    if (!options.algorithm) {
      throw new KryptosError("Algorithm is required");
    }
    if (!options.type) {
      throw new KryptosError("Type is required");
    }
    if (!options.use) {
      throw new KryptosError("Use is required");
    }

    return new Kryptos(options);
  }

  // private make

  private static get encryption(): Encryption {
    return {
      ec: this.makeEcEnc,
      oct: this.makeOctEnc,
      okp: this.makeOkpEnc,
      rsa: this.makeRsaEnc,
    };
  }

  private static get signature(): Signature {
    return {
      ec: this.makeEcSig,
      oct: this.makeOctSig,
      okp: this.makeOkpSig,
      rsa: this.makeRsaSig,
    };
  }

  private static makeAuto(options: KryptosAuto): IKryptos {
    const config = autoMakeConfig(options.algorithm);
    const operations = calculateKeyOps(config.use);

    const make: KryptosMake = {
      ...config,
      ...options,
      operations,
    };

    return new Kryptos({
      ...make,
      ...makeKey(make),
    });
  }

  private static makeEcEnc(options: KryptosMakeEcEnc): IKryptos {
    const make: KryptosMake = {
      ...options,
      type: "EC",
      use: "enc",
    };
    return KryptosKit.makeKryptos(make);
  }

  private static makeEcSig(options: KryptosMakeEcSig): IKryptos {
    const make: KryptosMake = {
      ...options,
      type: "EC",
      use: "sig",
    };
    return KryptosKit.makeKryptos(make);
  }

  private static makeOctEnc(options: KryptosMakeOctEnc): IKryptos {
    const make: KryptosMake = {
      ...options,
      type: "oct",
      use: "enc",
    };
    return KryptosKit.makeKryptos(make);
  }

  private static makeOctSig(options: KryptosMakeOctSig): IKryptos {
    const make: KryptosMake = {
      ...options,
      type: "oct",
      use: "sig",
    };
    return KryptosKit.makeKryptos(make);
  }

  private static makeOkpEnc(options: KryptosMakeOkpEnc): IKryptos {
    const make: KryptosMake = {
      ...options,
      type: "OKP",
      use: "enc",
    };
    return KryptosKit.makeKryptos(make);
  }

  private static makeOkpSig(options: KryptosMakeOkpSig): IKryptos {
    const make: KryptosMake = {
      ...options,
      type: "OKP",
      use: "sig",
    };
    return KryptosKit.makeKryptos(make);
  }

  private static makeRsaEnc(options: KryptosMakeRsaEnc): IKryptos {
    const make: KryptosMake = {
      ...options,
      type: "RSA",
      use: "enc",
    };
    return KryptosKit.makeKryptos(make);
  }

  private static makeRsaSig(options: KryptosMakeRsaSig): IKryptos {
    const make: KryptosMake = {
      ...options,
      type: "RSA",
      use: "sig",
    };
    return KryptosKit.makeKryptos(make);
  }

  private static makeKryptos(make: KryptosMake): IKryptos {
    return new Kryptos({
      ...make,
      operations: make.operations ?? calculateKeyOps(make.use),
      ...makeKey(make),
    });
  }
}
