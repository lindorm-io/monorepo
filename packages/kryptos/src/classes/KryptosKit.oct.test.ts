import MockDate from "mockdate";
import {
  TEST_OCT_KEY_B64,
  TEST_OCT_KEY_JWK,
  TEST_OCT_KEY_PEM,
  TEST_OCT_KEY_UTF,
} from "../__fixtures__/oct-keys";
import { KryptosKit } from "./KryptosKit";

const MockedDate = new Date("2024-01-01T08:00:00.000Z");
MockDate.set(MockedDate.toISOString());

jest.mock("crypto", () => ({
  ...jest.requireActual("crypto"),
  randomUUID: jest.fn().mockReturnValue("6e6f84b0-e125-5e3f-90ae-c65269668d98"),
}));

describe("KryptosKit (oct)", () => {
  describe("clone", () => {
    test("should clone", () => {
      const kryptos = KryptosKit.from.auto(TEST_OCT_KEY_B64);
      const cloned = KryptosKit.clone(kryptos);

      expect(cloned.toJSON()).toEqual(kryptos.toJSON());
      expect(cloned.toJWK()).toEqual(kryptos.toJWK());
      expect(cloned.export("b64")).toEqual(kryptos.export("b64"));
    });
  });

  describe("from", () => {
    describe("auto", () => {
      test("b64", () => {
        const kryptos = KryptosKit.from.auto(TEST_OCT_KEY_B64);

        expect(kryptos.toJSON()).toMatchSnapshot();
        expect(kryptos.toJWK()).toMatchSnapshot();
        expect(kryptos.export("b64")).toMatchSnapshot();
      });

      test("jwk", () => {
        const kryptos = KryptosKit.from.auto(TEST_OCT_KEY_JWK);

        expect(kryptos.toJSON()).toMatchSnapshot();
        expect(kryptos.toJWK()).toMatchSnapshot();
        expect(kryptos.export("jwk")).toMatchSnapshot();
      });

      test("pem", () => {
        const kryptos = KryptosKit.from.auto(TEST_OCT_KEY_PEM);

        expect(kryptos.toJSON()).toMatchSnapshot();
        expect(kryptos.toJWK()).toMatchSnapshot();
        expect(kryptos.export("pem")).toMatchSnapshot();
      });
    });

    test("b64", () => {
      const kryptos = KryptosKit.from.b64(TEST_OCT_KEY_B64);

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.toJWK()).toMatchSnapshot();
      expect(kryptos.export("b64")).toMatchSnapshot();
    });

    test("jwk", () => {
      const kryptos = KryptosKit.from.jwk(TEST_OCT_KEY_JWK);

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.toJWK()).toMatchSnapshot();
      expect(kryptos.export("jwk")).toMatchSnapshot();
    });

    test("pem", () => {
      const kryptos = KryptosKit.from.pem(TEST_OCT_KEY_PEM);

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.toJWK()).toMatchSnapshot();
      expect(kryptos.export("pem")).toMatchSnapshot();
    });

    test("utf", () => {
      const kryptos = KryptosKit.from.utf(TEST_OCT_KEY_UTF);

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.toJWK()).toMatchSnapshot();
      expect(kryptos.export("b64")).toMatchSnapshot();
    });
  });

  describe("is", () => {
    test("isEc", () => {
      const kryptos = KryptosKit.from.b64(TEST_OCT_KEY_B64);

      expect(KryptosKit.isEc(kryptos)).toBe(false);
    });

    test("isOct", () => {
      const kryptos = KryptosKit.from.b64(TEST_OCT_KEY_B64);

      expect(KryptosKit.isOct(kryptos)).toBe(true);
    });

    test("isOkp", () => {
      const kryptos = KryptosKit.from.b64(TEST_OCT_KEY_B64);

      expect(KryptosKit.isOkp(kryptos)).toBe(false);
    });

    test("isRsa", () => {
      const kryptos = KryptosKit.from.b64(TEST_OCT_KEY_B64);

      expect(KryptosKit.isRsa(kryptos)).toBe(false);
    });
  });

  describe("make", () => {
    test("auto", () => {
      const kryptos = KryptosKit.make.auto({
        algorithm: "HS384",
      });

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.export("b64")).toEqual({
        algorithm: "HS384",
        privateKey: expect.any(String),
        publicKey: "",
        type: "oct",
        use: "sig",
      });
    });

    test("enc", () => {
      const kryptos = KryptosKit.make.enc.oct({
        algorithm: "dir",
        encryption: "A256GCM",
      });

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.export("b64")).toEqual({
        algorithm: "dir",
        privateKey: expect.any(String),
        publicKey: "",
        type: "oct",
        use: "enc",
      });
    });

    test("sig", () => {
      const kryptos = KryptosKit.make.sig.oct({
        algorithm: "HS512",
      });

      expect(kryptos.toJSON()).toMatchSnapshot();
      expect(kryptos.export("b64")).toEqual({
        algorithm: "HS512",
        privateKey: expect.any(String),
        publicKey: "",
        type: "oct",
        use: "sig",
      });
    });
  });
});
