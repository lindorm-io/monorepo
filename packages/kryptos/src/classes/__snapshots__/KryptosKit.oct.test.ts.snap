// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KryptosKit (oct) from auto b64 1`] = `
{
  "algorithm": "HS512",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from auto b64 2`] = `
{
  "alg": "HS512",
  "iat": 1704096000,
  "k": "",
  "key_ops": [],
  "kid": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "kty": "oct",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from auto b64 3`] = `
{
  "algorithm": "HS512",
  "privateKey": "diYnyceZxmn18xGjVobBEwOSj2QOavHfv_tWGNuBpjND572Pa3qD8PDqDSrvoLtLOWyHdQ5lsmsuEIDcPgbPKp92HfNkawbKpCsVNBpoTlbZ-5jewLMREoGje9_pQzGSPLgh-cAkwtcrLUJNbwbyMGMlXmIJXeGukWsD6BfOAimNzPIyLf8QYMJYL9tzf16X4mQ1SvU76Y8Mqop6wz8ylAET7xWTivI-iOK8Zk1MiiomJww5w47Uz7X6Ha_uz7ctCESsyYMef9ZnYlsqwsHPrnP78ihyiv8cH7obubKJ6HkmsCnSTBOchDYxnmQiVZffuMSb8pScaIK6Vfef_1c7Vg",
  "publicKey": "",
  "type": "oct",
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from auto jwk 1`] = `
{
  "algorithm": "HS512",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": true,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from auto jwk 2`] = `
{
  "alg": "HS512",
  "iat": 1704096000,
  "k": "",
  "key_ops": [],
  "kid": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "kty": "oct",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from auto jwk 3`] = `
{
  "alg": "HS512",
  "k": "tt8sKlXAatelhbSo_IZgvyFzIiKPOKhv-IZTqAEkCkJZPVrO5PK9hb1NH9m9bJXa1nGmqlEVCjzsA1lwYsA2clgUvjQ3QlL3x7yDxzZqReg2x_W4XezFU2fBsGdsD2u-CW5FJ3WJZPsSDG5zOsARCdjvIyDiNO1Ryb-MRb0LWxcvsDdzNPIJNTf5adhf4Qsf5ZIWhvnSF5y9fglB6QRcskgwdRfGoYjjkiw3-h5UDrUnjzCPLQ7GVhq2gg_OyZJsduwTVYGObpAG6EYplvQZjNAWyekHivXfpc556jNTWCzfmAyZvvucsGIM5wuupicfn7AbA1a50fuppCyMwbBKSg",
  "kty": "oct",
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from auto pem 1`] = `
{
  "algorithm": "HS512",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from auto pem 2`] = `
{
  "alg": "HS512",
  "iat": 1704096000,
  "k": "",
  "key_ops": [],
  "kid": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "kty": "oct",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from auto pem 3`] = `
{
  "algorithm": "HS512",
  "privateKey": "-----BEGIN OCT PRIVATE KEY-----
ts4FwuyVlXJqw2ruyxbpzRpkknqiZicDS02J2YjfX+FwZ423Q42E0T/AKYkVXHNW
uD8Gi0aKMYytagkT7qJx3PsFCgDND01mXOg8pDhXv1n1Cbl8JxavKwbx87QCZ9Bm
PlUCPWDU8khZ/djxbtK/Hdl7ehFnk3KoFsEYphydCFhdI29UaURoj1ZNDV08Psnu
iqBCE9DgYc58QHaGRmpct9ORyYOJ4o+7jOD45cxeKZubVq4HIifiXk724DGP4M1/
crSKuYK1lddiniPwv/nnl0DAxI8/5Vz8NGXvGoJFaltQiZmKMC8y5O74DvIDknS3
cvQHzpn+t5VXn6eD+7A74Q==
-----END OCT PRIVATE KEY-----",
  "publicKey": "",
  "type": "oct",
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from b64 1`] = `
{
  "algorithm": "HS512",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from b64 2`] = `
{
  "alg": "HS512",
  "iat": 1704096000,
  "k": "",
  "key_ops": [],
  "kid": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "kty": "oct",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from b64 3`] = `
{
  "algorithm": "HS512",
  "privateKey": "diYnyceZxmn18xGjVobBEwOSj2QOavHfv_tWGNuBpjND572Pa3qD8PDqDSrvoLtLOWyHdQ5lsmsuEIDcPgbPKp92HfNkawbKpCsVNBpoTlbZ-5jewLMREoGje9_pQzGSPLgh-cAkwtcrLUJNbwbyMGMlXmIJXeGukWsD6BfOAimNzPIyLf8QYMJYL9tzf16X4mQ1SvU76Y8Mqop6wz8ylAET7xWTivI-iOK8Zk1MiiomJww5w47Uz7X6Ha_uz7ctCESsyYMef9ZnYlsqwsHPrnP78ihyiv8cH7obubKJ6HkmsCnSTBOchDYxnmQiVZffuMSb8pScaIK6Vfef_1c7Vg",
  "publicKey": "",
  "type": "oct",
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from jwk 1`] = `
{
  "algorithm": "HS512",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": true,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from jwk 2`] = `
{
  "alg": "HS512",
  "iat": 1704096000,
  "k": "",
  "key_ops": [],
  "kid": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "kty": "oct",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from jwk 3`] = `
{
  "alg": "HS512",
  "k": "tt8sKlXAatelhbSo_IZgvyFzIiKPOKhv-IZTqAEkCkJZPVrO5PK9hb1NH9m9bJXa1nGmqlEVCjzsA1lwYsA2clgUvjQ3QlL3x7yDxzZqReg2x_W4XezFU2fBsGdsD2u-CW5FJ3WJZPsSDG5zOsARCdjvIyDiNO1Ryb-MRb0LWxcvsDdzNPIJNTf5adhf4Qsf5ZIWhvnSF5y9fglB6QRcskgwdRfGoYjjkiw3-h5UDrUnjzCPLQ7GVhq2gg_OyZJsduwTVYGObpAG6EYplvQZjNAWyekHivXfpc556jNTWCzfmAyZvvucsGIM5wuupicfn7AbA1a50fuppCyMwbBKSg",
  "kty": "oct",
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from pem 1`] = `
{
  "algorithm": "HS512",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from pem 2`] = `
{
  "alg": "HS512",
  "iat": 1704096000,
  "k": "",
  "key_ops": [],
  "kid": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "kty": "oct",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from pem 3`] = `
{
  "algorithm": "HS512",
  "privateKey": "-----BEGIN OCT PRIVATE KEY-----
ts4FwuyVlXJqw2ruyxbpzRpkknqiZicDS02J2YjfX+FwZ423Q42E0T/AKYkVXHNW
uD8Gi0aKMYytagkT7qJx3PsFCgDND01mXOg8pDhXv1n1Cbl8JxavKwbx87QCZ9Bm
PlUCPWDU8khZ/djxbtK/Hdl7ehFnk3KoFsEYphydCFhdI29UaURoj1ZNDV08Psnu
iqBCE9DgYc58QHaGRmpct9ORyYOJ4o+7jOD45cxeKZubVq4HIifiXk724DGP4M1/
crSKuYK1lddiniPwv/nnl0DAxI8/5Vz8NGXvGoJFaltQiZmKMC8y5O74DvIDknS3
cvQHzpn+t5VXn6eD+7A74Q==
-----END OCT PRIVATE KEY-----",
  "publicKey": "",
  "type": "oct",
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from utf 1`] = `
{
  "algorithm": "HS512",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from utf 2`] = `
{
  "alg": "HS512",
  "iat": 1704096000,
  "k": "",
  "key_ops": [],
  "kid": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "kty": "oct",
  "nbf": 1704096000,
  "uat": 1704096000,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) from utf 3`] = `
{
  "algorithm": "HS512",
  "privateKey": "I-KCrCEuYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXouQUJDREVGR0hJSktMTU5PUFFSU1RVVldYWVouMDEyMzQ1Njc4OQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
  "publicKey": "",
  "type": "oct",
  "use": "sig",
}
`;

exports[`KryptosKit (oct) make auto 1`] = `
{
  "algorithm": "HS384",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "sign",
    "verify",
  ],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;

exports[`KryptosKit (oct) make enc 1`] = `
{
  "algorithm": "dir",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "encryption": "A256GCM",
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "encrypt",
    "decrypt",
  ],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "enc",
}
`;

exports[`KryptosKit (oct) make sig 1`] = `
{
  "algorithm": "HS512",
  "createdAt": 2024-01-01T08:00:00.000Z,
  "expiresIn": -1,
  "hasPrivateKey": true,
  "hasPublicKey": false,
  "id": "6e6f84b0-e125-5e3f-90ae-c65269668d98",
  "isActive": true,
  "isExpired": false,
  "isExternal": false,
  "notBefore": 2024-01-01T08:00:00.000Z,
  "operations": [
    "sign",
    "verify",
  ],
  "type": "oct",
  "updatedAt": 2024-01-01T08:00:00.000Z,
  "use": "sig",
}
`;
